bedita-app/app_controller.php

```diff
@@ -1304,14 +1304,39 @@ public function addPermissionsObjects()
     {
         if (!empty($this->params['form']['objects_selected'])) {
             $objectIds = $this->params['form']['objects_selected'];
+            /**
+             * [objectID][permissionName]
+             */
+            $duplicatePermissions = array();
             $this->Transaction->begin();
 
-            foreach ($objectIds as $id) {
+            foreach ($objectIds as $objectId) {
                 $permission = ClassRegistry::init('Permission');
-                $permission->add($id, $this->data['Permission']);
+                
+                foreach ($this->data['Permission'] as $permissionData) {
+                    if ($permission->isPermissionSet($objectId, $permissionData)) {
+                        if (!array_key_exists($objectId, $duplicatePermissions)) {
+                            $duplicatePermissions[$objectId] = array();
+                        }
+                        
+                        $duplicatePermissions[$objectId][] = $permissionData['name'];
+                    } else {
+                        $permission->add($objectId, array($permissionData));
+                    }
+                }
             }
 
             $this->Transaction->commit();
```

Comment:
If `$permission->add()` fails all transaction is rollback.
Maybe we should remove the transaction here.

`Permission::isPermissionSet()` check just the `permissions.flag`, so you should check againts `flag` and `ugid` (`group_id` related to group name you have in the data array) and `'switch' => 'group'`.

Something as:

```php
$Permission = ClassRegistry::init('Permission');
$Group = ClassRegistry::init('Group');

foreach ($objectIds as $objectId) {
    foreach ($this->data['Permission'] as $permissionData) {
        $ugid = $group->field('id', array('name' => $permissionData['name']));
        $countPerms = $Permission->find('count', array(
            'conditions' => array(
                'switch' => 'group',
                'ugid' => $ugid,
                'flag' => $permissionData['flag'],
            ),
        ));
        if ($countPerms > 0) {
            $duplicatePermissions[$objectId][] = $permissionData['name']; // don't need check on key exists
            continue;
        }
        
        try {
            $Permission->add($objectId, array($permissionData));
        } catch (Exception $e) {
            // trace errros for view???
            $this->log('Error saving permission'); // more info is better
        }
    }
}
```

Reply:
I think we can leave the transaction, It's ok to rollback if It fails for any other reason.

Thanks for your snippet, I'll change the code. I guess I should add `object_id => $objectId` to the `$Permission->find(...)` conditions, right?

Reply:
correct, the condition on `object_id` is missing